### DISCUSSION TITLE
[SvgDecoder][getDstSize] dstWidth and dstHeight are not translated to px

### DISCUSSION THREAD

**Discussion by MakarenkoVlad** [coil-kt/coil#1455] [General]
<img width="944" alt="image" src="https://user-images.githubusercontent.com/53937436/189608357-9534e43e-1476-4d5a-b171-e6d2a9c857fc.png">

# Main point here is to load icon with its svg size. But, we have only few options to load picture: scaled to icon bounds, with predefined size, with original size in px. But we want this px translated to dp and again to px for its equal look on different devices.
## _Please, do not close this issue without discussion_

For original size dstWidth and dstHeight are not translated to pixels before returning from getDstSize method. That leads to smaller image size which is not correct.
Prerequisites:

Steps to reproduce:
Implement such code for image download
```
val CONTENT_SCALE = ContentScale.None

val painter = rememberAsyncImagePainter(
        model = ImageRequest.Builder(LocalContext.current)
            .data("https://svgshare.com/i/mX9.svg")
            .size(Size.ORIGINAL)
            .build(),
        imageLoader = ImageLoader.Builder(LocalContext.current)
            .components {
                add(SvgDecoder.Factory())
            }
            .build(),
        error = fallbackPainter,
        fallback = fallbackPainter,
        contentScale = CONTENT_SCALE
    )

Box(
        Modifier
            .sizeIn(maxWidth = 40.dp, maxHeight = 40.dp)
            .toolingGraphicsLayer()
            .paint(
                painter,
                colorFilter = colorFilter,
                contentScale = CONTENT_SCALE,
                alignment = Alignment.Center,
            )
    )
```

Problem place:
```
private fun getDstSize(srcWidth: Float, srcHeight: Float, scale: Scale): Pair<Float, Float> {
        if (options.size.isOriginal) {
            val dstWidth = if (srcWidth > 0) srcWidth else DEFAULT_SIZE
            val dstHeight = if (srcHeight > 0) srcHeight else DEFAULT_SIZE
            return dstWidth to dstHeight
        } else {
            val (dstWidth, dstHeight) = options.size
            return dstWidth.toPx(scale) to dstHeight.toPx(scale)
        }
    }
```

<img width="367" alt="Screenshot 2022-09-09 at 12 24 54" src="https://user-images.githubusercontent.com/53937436/189318193-80e114c6-39cf-4679-a246-0f32f0def7cb.png">

With compose density it could be implemented like this
Possible fix:
```

private fun getDstSize(srcWidth: Float, srcHeight: Float, scale: Scale): Pair<Float, Float> {
        if (options.size.isOriginal) {
            val dstWidth = if (srcWidth > 0) srcWidth else DEFAULT_SIZE
            val dstHeight = if (srcHeight > 0) srcHeight else DEFAULT_SIZE
            return density.run { dstWidth.dp.toPx() to dstHeight.dp.toPx() }
        } else {
            val (dstWidth, dstHeight) = options.size
            return dstWidth.toPx(scale) to dstHeight.toPx(scale)
        }
    }
```
<img width="369" alt="Screenshot 2022-09-09 at 12 26 25" src="https://user-images.githubusercontent.com/53937436/189318366-fd06ec73-eba5-4b18-b908-1a86efda67c2.png">

**Comment by MakarenkoVlad** [comment]
Maybe you can add a new method for ImageRequest.Builder().scale that can take a ratio?

**Reply by colinrtwhite** [reply]
The way Coil's size system at the moment is built, it's not possible to implement an `ImageRequest.Builder.size` method that accepts a ratio.

**Comment by colinrtwhite** [comment]
Hmm for something custom like this I think you might have to create a custom version of `SvgDecoder`. You could also potentially rewrite the SVG using Okio before it reaches the SVG decoder to replace its view box with one scaled to the device's density.

**Reply by MakarenkoVlad** [reply]
I'll do as u said. Ty)))
